In [116]:
import pandas as pd
import numpy as np

In [117]:
file = 'autos.csv'
autos = pd.read_csv (file, encoding='Windows-1252')

In [118]:
print (autos.info())
print (autos.head(5))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [119]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [120]:
ren_columns={'yearOfRegistration' : 'registration_year',
'monthOfRegistration' : 'registration_month',
'notRepairedDamage' : 'unrepaired_damage',
'dateCreated' : 'ad_created'}
autos.rename(columns=ren_columns, inplace=True)

In [121]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [122]:
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-11 22:38:16  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  registration_year  gearbox       powerPS  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuelType       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [123]:
autos.rename(columns={'odometer' : 'odometer_km'}, inplace=True)

In [124]:
autos['price'] = autos['price'].str.extract('(\d+)', expand=False).astype(float)
autos['odometer_km'] = autos['odometer_km'].str.extract('(\d+)', expand=False).astype(float)

In [125]:
autos.dtypes

dateCrawled            object
name                   object
seller                 object
offerType              object
price                 float64
abtest                 object
vehicleType            object
registration_year       int64
gearbox                object
powerPS                 int64
model                  object
odometer_km           float64
registration_month      int64
fuelType               object
brand                  object
unrepaired_damage      object
ad_created             object
nrOfPictures            int64
postalCode              int64
lastSeen               object
dtype: object

In [126]:
price = autos['price']
speed = autos['odometer_km']

In [127]:
price.unique(), speed.unique()

(array([  5.,   8.,   4.,   1.,   7., 300., 250., 590., 999., 350.,   3.,
         18.,  41.,  25.,  48.,  90., 777.,   0.,  80.,   2.,  13.,  11.,
        900.,   9.,  56.,   6., 650., 700., 995.,  40.,  10.,  31., 449.,
        399.,  34.,  36., 800., 600., 320.,  12.,  20., 150., 550., 190.,
        200.,  32.,  29., 799., 500.,  23., 499.,  50.,  15.,  21., 890.,
         14., 450., 850.,  17.,  24., 888., 750.,  26.,  16.,  22.,  19.,
        230., 400., 580., 649., 730., 760., 950., 299.,  42., 790., 699.,
         33., 599., 690., 100.,  52., 380., 180., 490.,  46., 333., 170.,
        940., 990., 375., 270., 149., 980.,  70., 280., 899., 949.,  28.,
         35.,  38., 870., 689., 460., 290., 549., 830., 480.,  27., 370.,
        625., 220.,  69.,  99., 749.,  73.,  39.,  37.,  54., 570., 225.,
        660., 555.,  75., 199.,  30., 129.,  65., 430.,  66., 475., 330.,
        675., 325., 137., 269., 530., 360.,  93., 669., 120.,  74., 998.,
        996., 720., 849.,  62.,  53., 

In [128]:
# price.max

In [129]:
# removing the outliers
autos = autos[price.between(price.min(axis=0),price.max(axis=0), inclusive=False)]
autos.reset_index(inplace = True)
autos = autos[speed.between(speed.min(axis=0),speed.max(axis=0), inclusive=False)]

/dataquest/system/env/python3/lib/python3.4/site-packages/ipykernel/__main__.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.


In [130]:
autos.shape

(16039, 21)

In [131]:
dates = ['dateCrawled', 'ad_created', 'lastSeen']

for i in dates:
    print(autos[i].value_counts(normalize=True, dropna=False))

2016-03-21 19:53:48    0.000125
2016-03-23 22:53:33    0.000125
2016-03-27 20:48:00    0.000125
2016-04-04 17:50:45    0.000125
2016-04-01 22:55:00    0.000125
2016-03-07 19:53:43    0.000125
2016-03-23 16:45:54    0.000125
2016-03-20 20:50:18    0.000125
2016-03-26 17:49:36    0.000125
2016-03-13 18:36:42    0.000125
2016-03-30 11:54:33    0.000125
2016-03-20 14:39:11    0.000125
2016-03-11 20:56:02    0.000125
2016-03-23 20:36:18    0.000125
2016-03-22 17:36:39    0.000125
2016-04-02 15:47:03    0.000125
2016-03-21 14:06:23    0.000125
2016-03-19 12:54:33    0.000125
2016-03-08 14:39:13    0.000125
2016-03-16 17:58:35    0.000125
2016-03-24 22:51:23    0.000125
2016-03-16 13:49:52    0.000125
2016-03-24 13:50:39    0.000125
2016-04-03 18:40:27    0.000125
2016-03-28 22:36:20    0.000125
2016-03-20 17:47:21    0.000125
2016-03-31 13:06:20    0.000125
2016-04-03 15:58:44    0.000125
2016-04-02 12:52:59    0.000125
2016-03-29 21:55:05    0.000125
                         ...   
2016-03-

In [132]:
autos['registration_year'].describe()

count    16039.000000
mean      2003.959162
std         42.673973
min       1111.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       5911.000000
Name: registration_year, dtype: float64

likely using 9999 as unknown (max)
likely using 1000 as unknown (min)

In [133]:
autos['registration_year'].value_counts(normalize=True)

2000    0.064218
2005    0.061039
1999    0.059168
2003    0.057734
2004    0.055801
2006    0.055614
2001    0.053682
2002    0.050377
1998    0.049816
2007    0.047634
2008    0.046013
2009    0.043082
1997    0.038593
2011    0.032546
2010    0.032546
2017    0.029428
1996    0.029241
2012    0.026061
1995    0.025189
2016    0.024877
2013    0.016834
2014    0.014278
1994    0.013405
2018    0.008853
1993    0.008666
2015    0.008168
1992    0.007669
1991    0.006422
1990    0.006172
1989    0.002743
          ...   
1977    0.000499
1969    0.000499
1973    0.000436
1968    0.000436
1975    0.000436
1974    0.000436
1965    0.000374
1960    0.000374
1976    0.000249
1964    0.000249
1963    0.000187
1937    0.000125
1956    0.000125
1910    0.000125
1959    0.000125
4800    0.000062
1951    0.000062
1943    0.000062
1111    0.000062
1958    0.000062
1934    0.000062
4100    0.000062
1938    0.000062
1929    0.000062
1941    0.000062
5911    0.000062
1948    0.000062
1955    0.0000

In [134]:
registration = autos['registration_year']
autos = autos[registration.between(registration.min(axis=0),registration.max(axis=0))]
autos.reset_index(inplace=True)

In [137]:
registration.describe()

count    16039.000000
mean      2003.959162
std         42.673973
min       1111.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       5911.000000
Name: registration_year, dtype: float64